### DISCUSSION TITLE
Rust crates with regexmanager don't get updated

### DISCUSSION THREAD

**Discussion by davidkna** [renovatebot/renovate#17439] [Request Help] [answered]
Sample repository: https://github.com/davidkna/renovate-cargo-install

The following renovate config doesn't seem to update the binary crates that get installed in the workflow file (`run: cargo install --version 0.4.7 cargo-llvm-cov`), despite [successfully picking up the current version](https://github.com/davidkna/renovate-cargo-install/issues/1):
```json
{
    "$schema": "https://docs.renovatebot.com/renovate-schema.json",
    "extends": ["config:base"],
    "regexManagers": [{
      "fileMatch": ["^(workflow-templates|\\.github\\/workflows)\\/[^/]+\\.ya?ml$"],
      "matchStrings": ["cargo install.*--version (?<currentValue>\\S+).*\\s(?<depName>\\S+)"],
      "datasourceTemplate": "crate"
    }]
 }
```

Partial log output:
```
DEBUG: packageFiles with updates
{
  "baseBranch": "main",
  "config": {
    "regex": [
      {
        "packageFile": ".github/workflows/workflow.yml",
        "deps": [
          {
            "depName": "cargo-llvm-cov",
            "currentValue": "0.4.7",
            "datasource": "crate",
            "replaceString": "cargo install --version 0.4.7 cargo-llvm-cov",
            "depIndex": 0,
            "updates": [],
            "warnings": [],
            "versioning": "cargo",
            "sourceUrl": "https://github.com/taiki-e/cargo-llvm-cov",
            "dependencyUrl": "https://crates.io/crates/cargo-llvm-cov",
            "currentVersion": "0.4.7"
          }
        ],
        "matchStrings": [
          "cargo install.*--version (?<currentValue>\\S+).*\\s(?<depName>\\S+)"
        ],
        "datasourceTemplate": "crate"
      }
    ]
  }
}
```

Is this a bug in renovate, or is there something wrong with my config?

**Comment by Gabriel-Ladzaretti** [comment]
I've debugged this, and it seems that all valid release updates are being filtered out as there is no `lockedVersion` as you are using a `regexManger` and obviously there is not lock file present.

![image](https://user-images.githubusercontent.com/97394622/187077997-e5e125de-419c-434f-bf80-75cd435e6271.png)

https://github.com/renovatebot/renovate/blob/dafda2e597eab338df5560dad0dbe7ada14adae1/lib/workers/repository/process/lookup/index.ts#L266-L268

...
https://github.com/renovatebot/renovate/blob/dafda2e597eab338df5560dad0dbe7ada14adae1/lib/workers/repository/process/lookup/index.ts#L300-L303

@viceice WDYT?

**Reply by viceice** [reply]
it's probably the versioning. major is zero, so renovate won't update to next minor. locked version isn't required for PR.

**Reply by Gabriel-Ladzaretti** [reply]
problem is that the update array is empty as all valid releases are being filtered out
![image](https://user-images.githubusercontent.com/97394622/187078848-9ccb2e11-5547-4620-9156-b183a37fb07b.png)

edit:
there is only non-major bucket - 
![image](https://user-images.githubusercontent.com/97394622/187079030-960c2740-297b-4cb5-bdef-b2ab273845c9.png)

**Reply by viceice** [reply]
as i said, it's a versioning issue. checkout cargo versioning. by default separate minor and patch is disabled, so that's why there is only a non major bucket. 

cargo uses semver 2.0, so our renovate cargo extend npm versioning.

when major is zero, minor is change is like a unstable update, which is suppressed by ignore unstable setting.

**Reply by Gabriel-Ladzaretti** [reply]
You are probably right, seems disabling `ignoreUnstable` doesn't help here either

```
DEBUG: Checking for config file in C:\projects\ladzaretti-renovate/config.js
 WARN: Config needs migrating
       "originalConfig": {
         ....
         "ignoreUnstable": false,

...
"config": {
         "regex": [
           {
             "packageFile": ".github/workflows/workflow.yml",
             "deps": [
               {
                 ...
                 "updates": [],
                 "warnings": [],
                 "versioning": "cargo",
                 "sourceUrl": "https://github.com/taiki-e/cargo-llvm-cov",
                 "dependencyUrl": "https://crates.io/crates/cargo-llvm-cov",
                 "currentVersion": "0.4.7"
               }
             ],
```

**Comment by williamboman** [comment, marked as answer]
I got somewhat stuck at this as well. I use a custom `regexManager` to update crates (not expressed in a `Cargo.toml` file but in a custom file format). Reading the [README](https://github.com/renovatebot/renovate/tree/main/lib/modules/versioning/cargo) of the `cargo` versioning strategy helped me realize it rightfully treats versions according to the cargo spec: 

> In Cargo, 1.2.3 doesn't mean "exactly 1.2.3", it actually means >=1.2.3 <2.0.0. So this is like the equivalent of ^1.2.3 in npm.

Changing `versioning` to `semver` for the regex manager made it behave as I'd expect (treats versions as fixed):

```json5
{
  regexManagers: [
    {
      fileMatch: "package\\.yaml$",
      matchStrings: [
        "pkg:cargo\\/(?<depName>.+)@(?<currentValue>.+)"
      ],
      datasourceTemplate: "crate",
      // In cargo versioning, "1.2.3" doesn't mean "exactly 1.2.3" but rather ">=1.2.3 <2.0.0".
      // We on the other hand do want to treat versions as fixed - so we change versioning strategy.
      versioningTemplate: "semver",
    },
  ]
}
```

**Reply by gillyobeast** [reply]
one can also use `versioning` in a `packageRules` section to achieve the same thing, if you're using the regex manager to update from multiple datasources and don't want to override the versioning for other datasources:
```json5
{
      // When upgrading tool versions from cargo in .env with regex, use semver instead of cargo's versioning.
      matchManagers: [
        "regex"
      ],
      matchDatasources: [
        "crate"
      ],
      versioning: "semver"
    }
```
